### PR TITLE
Release - 1.2.7

### PR DESCRIPTION
**Pull request by ryanio**
## Description

This PR introduces `web3.js` version `1.2.7` beginning with `1.2.7-rc.0`.

[Please see the release notes](https://github.com/ethereum/web3.js/releases/tag/1.2.7-rc.0) for more.

### Added

- Add revert reason support to sendSignedTransaction (#3345)
- ENS module extended with the possibility to add a custom registry (#3301)
- Missing ENS Registry methods and Resolver.supportsInterface method added (#3325)
- Add optional gas type to AbiItem typescript definitions (for ABIs generated by Vyper) (#3437)
- Add auto-reconnect option for Websockets (#3092, #1085, #1391, #1558, #1852, #1646)

### Changed

- Ensure '0x' prefix is existing for Accounts.sign and Accounts.privateKeyToAccount (#3041)
- Repository cleanup (#3443)
  - Removed old `docs/_build` folder
  - Removed old bower and meteor artifacts
  - Moved logo assets to own folder
  - Moved github assets to own folder
- Remove @types/node from (non-dev) dependency tree (#3965, #3227)

### Fixed

- Add missing subscription.on('connected') TS type definition (#3319)
- Add missing bignumber.js dependency for TS types (#3386)
- Upgrade swarm-js to 0.1.40 to remove npm vulnerability warning (#3399)
- Upgrade devDeps to resolve security warnings (#3464)
  - dtslint 0.4.2 => 3.4.1
  - definitelytyped-header-parser 1.0.1 => 3.9.0
- Race-condition when subscribing to historical logs as first client request (#3389)
- Fix crash when using Web-Workers by removing any-promise dependency (#3377 #2211 #1774)
- MaxListenersExceededWarning event emitter warning mitigated (#1648)

## Compare View

[v1.2.6 -> v1.2.7](https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0) 

## Type of change

- [x] Release

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [ ] I ran ```npm run test:unit``` with success.
- [ ] I have executed ``npm run test:cov`` and my test cases do cover all lines and branches of the added code.
- [ ] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [x] I have tested my code on the live network.

### CODE REVIEW

**Comment by ryanio** [comment]
Hey everyone, @cgewecke and I are happy to announce release version `1.2.7`.

We plan to leave this PR open for one week (until next **Tuesday, April 21**) for community feedback and reviewer approvals. If there are any large changes requested, they will be published and released as incremented `rc` releases.

@michaelsbradleyjr @alcuadrado @gnidan hey 👋 please let us know if you could take a 👀 and give us your 👍 for this release. Thank you!

**Comment by cgewecke** [comment]
As part of the RC process, am doing some test installations in other projects and running their tests to see if everything.  
+ installs properly
+ builds
+ tests pass

This list will be actively edited: 

| Project | Desc | Status |
| ------ | ----- | ----- |
| [oz-test-helpers][1] | Test utils. Subset of their suite is web3 contracts | ✅  |
| [oceanprotocol/squid-js][2] | TS API ~95 tests + tsc + webpack build | Only used by sub-deps |
| [gnosis/dex-react][3] | TS React app ~200 tests + webpack build + jest | ✅  (Test/Build)|

[1]: https://github.com/OpenZeppelin/openzeppelin-test-helpers
[2]: https://github.com/oceanprotocol/squid-js
[3]: https://github.com/gnosis/dex-react

**Comment by nivida** [comment]
Is there a compare view existing to see the diff of 1.2.6 and 1.2.7 as in all other releases? also, the changelog we always had and which notified the issue/PR creators is missing in the description? can you reference the release guidelines as well? and is the RC version already released because I can't see the ``in-progress`` label? (I know some annoying questions) :-)

File: gulpfile.js [hunk 1]
```diff
@@ -119,6 +119,7 @@ var ugliyOptions = {
     }
 };
 
+// Apply lerna.json version to root package.json
```
Comment (nivida) [inline gulpfile.js:122]: I think these comments are non-related to this PR.. see point 4.1 of the contribution guidelines.

**Comment by ryanio** [comment]
> Is there a compare view existing to see the diff of 1.2.6 and 1.2.7 as in all other releases?

Yes thanks, just added!

> also, the changelog we always had and which notified the issue/PR creators is missing in the description? can you reference the release guidelines as well?

Will add now.

> and is the RC version already released because I can't see the `in-progress` label? (I know some annoying questions) :-)

Just added, thanks ;)

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
-    "name": "web3",
+    "name": "web3.js",
```
Comment (nivida) [inline package.json:2]: What is the reason to rename the root/dev package from web3 to web3.js? Also this is a non-required change :-)

File: package.json [hunk 3]
```diff
@@ -20,10 +20,11 @@
         }
     ],
     "scripts": {
+        "version": "npm run build-all",
```
Comment (nivida) [inline package.json:23]: Shouldn't those modified or additional commands not be a separate PR to be aligned with the contribution guidelines from @cgewecke?

File: package.json [hunk 4]
```diff
@@ -132,7 +133,7 @@
         "karma-firefox-launcher": "^1.2.0",
         "karma-mocha": "^1.3.0",
         "karma-spec-reporter": "0.0.32",
-        "lerna": "^3.18.3",
+        "lerna": "^3.20.2",
```
Comment (nivida) [inline package.json:136]: What exactly was the reason to bump two minor versions of lerna? 🤔

File: package.json (hunk 2 above)
Comment (ryanio) [reply package.json:2]: We were running into lerna conflicts/bugs in our release process because the root package name was identical to one of the subpackages, so this fixed that and simplified our latest [release steps](https://github.com/ethereum/web3.js/blob/81e6cd5bf25cb13c7081e8c770eebc2416726965/RELEASE.md#release-candidate-rc-release-procedure).

Comment (nivida) [reply package.json:2]: hmm.. never had that.. 🤔 (typically lerna).. it should ignore this root ``package.json`` at all especially because of the ``private: true``

Comment (michaelsbradleyjr) [reply package.json:2]: Probably okay, i.e. it's not uncommon to have the `"name"` be `"root"` in the root `package.json` file of a lerna monorepo.

Comment (michaelsbradleyjr) [reply package.json:2]: See e.g. https://github.com/embarklabs/embark/blob/master/package.json#L41

Comment (nivida) [reply package.json:2]: Yep, just commented it to have the reason documented. :-) @michaelsbradleyjr thanks!

File: package.json (hunk 3 above)
Comment (michaelsbradleyjr) [reply package.json:23]: ***NOT*** a good idea to have a script named the same as a discrete step in the NPM release-lifecycle unless you fully intend to follow the documented behavior (***not*** sure lerna can/will/do/acknowledge that re: the root `package.json` in a `lerna publish` scenario)...
> Run AFTER bumping the package version, but BEFORE commit.

See: https://docs.npmjs.com/misc/scripts

Maybe that's the case, but NOTE-ing out of caution.

Comment (ryanio) [reply package.json:23]: Yes thanks that was exactly our intention, it is to build the minified file after bumping package versions but before committing the working state.

File: package.json (hunk 4 above)
Comment (michaelsbradleyjr) [reply package.json:136]: The lerna project is quite good about adhering to semver, so this should only pick-up bug fixes and new featuresl

File: package.json (hunk 3 above)
Comment (michaelsbradleyjr) [reply package.json:23]: Hmmm... not entirely sure, but sounds like maybe `"files"` in `package.json` (in child packages) could be added / edited; and that `.gitignore` (root or children) could be adjusted to ignore `*.min.js`. Main point: if you're specifying a `"version"` script, something's probably not right.

It's been awhile since I looked closely at this monorepo, but if you have `.npmignore` in root or child packages, nuke with prejudice and prefer `"files"` in all child packages' `package.json` — it will save you a world of trouble.

**Comment by nivida** [comment]
> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-40a5d984375cfacc5cc552a01648ee51R20

Could we not add this disclaimer to the ``README.md`` and documentation and remove the in-line comment? This information is probably important for some old projects which are integrating web3.js idk.. (only the duplicated comments everywhere should have brought that up) :-)

I mean [this](https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-8c057c1e8f6102642a2309bb9fce3731R20) says 3.7.. do we now have 3.0 or 3.7 as min. version? :P

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-8c363e4016c0523ddfd8355b2b3dcf48L26

Did this probably also improve the ``Promise`` chaining functionality? We do have 1-2 issues about it here on GitHub. 

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-b9a9f37d5a3a90f3953a269ddca60a65R62

Typo it should be: `` // Subscription might not exist``

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-a383cea7a2a9d92a58631e2d44a94ffeR201

@cgewecke is there a way to document the timing issue we noticed? 🤔 

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-5ca1164e985980d3e045ad9298bbec26R9

Is this required or do I/We have a badly defined type? idk.. 

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-bd11c69500f2885e9e6ebfaf8fd2da13R26

I think this comment is I think from my side.. 😅 but we should probably create an issue out of it to track it idk... and [this](https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-6c94e090da4006f68ea1721d17f2a531R27) is a lie the generic ``Contract`` was/is always implemented.

```javascript
class MyContract extends eth.Contract {}
```

> https://github.com/ethereum/web3.js/compare/v1.2.6...v1.2.7-rc.0#diff-7fb0ce5099d9515831133d423fef6c75R30

Can we add the explanation of it to the changelog or so? otherwise, those addresses do look like cheeked in... 

> revert msg sendSignedTransaction

As I informed Chris already is the additional network request not required.. should we inform the user about it? or do you think as I'm doing that this isn't required for the time being?

Thanks for all of your great work on this release! I'm really happy to see my started improvements are getting done and improved! ❤️

File: package.json (hunk 3 above)
Comment (michaelsbradleyjr) [reply package.json:23]: Also, consider `"prepublishOnly'"` in `package.json` scripts.

Comment (nivida) [reply package.json:23]: > Hmmm... not entirely sure, but sounds like maybe "files" in package.json (in child packages) could be added / edited; and that .gitignore (root or children) could be adjusted to ignore *.min.js. Main point: if you're specifying a "version" script, something's probably not right.

I agree.. I've added the ``files``property to the ``package.json`` files in ``2.x``... 

> and that .gitignore (root or children) could be adjusted to ignore *.min.js. 

Sadly no.. I've noticed the min file has to be on GitHub because many do use it for their tutorials and courses or of whatever reason :-)

Comment (nivida) [reply package.json:23]: @ryanio 
> NOT a good idea to have a script named the same as a discrete step

I would add this command/alias to the required places without defining a ``version`` command because this is for some/many devs a "hidden logic" :D

File: package.json (hunk 4 above)
Comment (nivida) [reply package.json:136]: Yep, I agree and know.. just thought it is out of the scope of this PR idk (4.1 contribution rules) :-)

File: package.json (hunk 3 above)
Comment (cgewecke) [reply package.json:23]: There's more about this in the lerna documentation for [version command lifecycle scripts][1].

[1]: https://github.com/lerna/lerna/tree/master/commands/version#lifecycle-scripts

File: package.json (hunk 4 above)
Comment (cgewecke) [reply package.json:136]: For additional background: the existence of several moderate/high vulnerability audit warnings originating in Lerna was flagged by a longtime contributor in this [2744 comment, yesterday][1]. 

[1]: https://github.com/ethereum/web3.js/issues/2744#issuecomment-613590354

**Comment by cgewecke** [comment]
@nivida In your [comment][1] with detailed review suggestions, the links aren't resolving correctly. When I click on the diff hashes it shows "nothing to compare" page. Are you also seeing this on your side? 

The suggested change to sendSignedTransaction is being tracked in #3457. It's an efficiency improvement that won't change correctness. Unsure what we should "inform the user about" here - could you clarify? 

Unfortunately your suggestion about that came right as we agreed to freeze public-facing changes to prepare for a release so it will need to go in 1.2.8. 

[1]: https://github.com/ethereum/web3.js/pull/3470#issuecomment-613795673

**Comment by ryanio** [comment]
> I mean this says 3.7.. do we now have 3.0 or 3.7 as min. version? :P

dtslint was only checking TypeScript 3.9 prior to #3464. There were some type errors on earlier versions, which is why you see those minimum comments, it's actually to programmatically let dtslint know. I've been meaning to go back and see if I can resolve them since it's just a few of them out of all the packages, but haven't gotten a chance yet.

edit: resolved in https://github.com/ethereum/web3.js/pull/3479

**Comment by ryanio** [comment]
Just an update since today was our planned day for release.

@cgewecke and I still have a few review comments we are addressing from @nivida before considering ready so please give us some additional time to finish the aforementioned threads.

**Comment by cgewecke** [comment]
- [x] Release notes: remove comments about MaxListenersExceededWarning
- [x] Release notes: describe and link to auto-reconnect configuration in docs
  + https://web3js.readthedocs.io/en/v1.2.7-rc.0/web3-eth.html#configuration

**Review by cgewecke — APPROVED** [review]
Yes, great work @ryanio! 🎉

**Comment by holgerd77** [comment]
Congrats! 🐡🐠🐟

**Comment by paub444** [comment]
I'm facing an issue when I do require('web3') inside a worker. It works for the first call, but nodejs crashes on the second call to the program.   
Strangely, this issue happens when I use require('web3') in the main thread. I cannot avoid this because I need web3 in the main thread as well.

Is there an issue with updation of require cache by any chance?
